packages/syft/src/syft/service/worker/worker_pool_service.py

```diff
@@ -178,6 +179,109 @@ def delete_worker(
             message=f"Worker with id: {worker_id} deleted successfully from pool: {worker_pool.name}"
         )
 
+    @service_method(
+        path="worker_pool.get_worker",
+        name="get_worker",
+        roles=DATA_OWNER_ROLE_LEVEL,
+    )
+    def get_worker(
+        self, context: AuthedServiceContext, pool_name: str, worker_id: UID
+    ) -> Union[SyftWorker, SyftError]:
+        result = self.stash.get_by_name(context.credentials, pool_name=pool_name)
+        if result.is_err():
+            return SyftError(message=f"{result.err()}")
+        if result.ok() is None:
+            return SyftError(
+                message=f"Worker Pool with name: {pool_name} was not found"
+            )
+
+        worker_pool = result.ok()
+
+        worker = None
+        for w in worker_pool.workers:
+            if w.id == worker_id:
+                worker = w
+                break
+
+        if worker is None:
+            return SyftError(
+                message=f"Worker with id: {worker_id} not found in pool: {worker_pool.name}"
+            )
+
+        worker_status = self.get_worker_status(
+            context=context, pool_name=pool_name, worker_id=worker.id
+        )
+        if isinstance(worker_status, SyftError):
+            return worker_status
+        elif worker_status != WorkerStatus.PENDING:
+            for worker in worker_pool.workers:
+                if worker.id == worker_id:
+                    worker.status = worker_status
+                    break
+            result = self.stash.update(
+                credentials=context.credentials,
+                obj=worker_pool,
+            )
+            if result.is_err():
+                return SyftError(message=f"{result.err()}")
+            else:
+                return worker
+        else:
+            return worker
+
+    @service_method(
+        path="worker_pool.get_worker_status",
+        name="get_worker_status",
+        roles=DATA_OWNER_ROLE_LEVEL,
+    )
+    def get_worker_status(
+        self, context: AuthedServiceContext, pool_name: str, worker_id: UID
+    ) -> Union[WorkerStatus, SyftError]:
+        result = self.stash.get_by_name(context.credentials, pool_name=pool_name)
+        if result.is_err():
+            return SyftError(message=f"{result.err()}")
+
+        if result.ok() is None:
+            return SyftError(
+                message=f"Worker Pool with name: {pool_name} was not found"
+            )
+
+        worker_pool = result.ok()
+
+        worker = None
+        for w in worker_pool.workers:
+            if w.id == worker_id:
+                worker = w
+                break
+
+        if worker is None:
+            return SyftError(
+                message=f"Worker with id: {worker_id} not found in pool: {worker_pool.name}"
+            )
+        try:
+            client = docker.from_env()
+            worker_status = client.containers.get(worker.container_id).status
```

Comment:
MAYBE: We could create new utility function, named
_get_worker_container_status something like

```python3
def _get_worker_container_status(
    worker: SyftWorker, docker_client: Optional[docker.DockerClient] = None
) -> Union[WorkerStatus, SyftError]:
    container = _get_worker_container(worker, docker_client)
    if isinstance(container, SyftError):
        return container
    
    container_status = container.status
    syft_container_status = None
    if container_status == "running":
        syft_container_status = WorkerStatus.RUNNING
    elif container_status in ["paused", "removing", "exited", "dead"]:
        syft_container_status = WorkerStatus.STOPPED
    elif container_status == "restarting":
        syft_container_status = WorkerStatus.RESTARTED
    elif container_status == "created":
        syft_container_status = WorkerStatus.PENDING
    else:
        return SyftError(
            message=f"Unknown container status: {container_status}"
        )
    

    return syft_container_status
```

Reply:
and maybe we could call this inside get_worker?

Reply:
Incorporated this into latest commit.

Reply:
I feel like we could decouple this `_get_worker_container_status` a little bit. Currently it's doing too many things. Also to convert from `Container.status` (`str`) to `WorkerStatus` we can use a `dict` to look up, making it cleaner and a tiny bit faster. Something like this

```py
_CONTAINER_STATUS_TO_WORKER_STATUS: dict[str, WorkerStatus] = dict(
    [
        ("running", WorkerStatus.RUNNING),
        *(
            (status, WorkerStatus.STOPPED)
            for status in ["paused", "removing", "exited", "dead"]
        ),
        ("restarting", WorkerStatus.RESTARTED),
        ("created", WorkerStatus.PENDING),
    ]
)
```

Inside `def get_worker` could just be

```py
container = _get_worker_container(worker)
if isinstance(container, SyftError):
    return container

container_status = container.status
worker_status = _CONTAINER_STATUS_TO_WORKER_STATUS.get(container_status)
if worker_status is None:
    return SyftError(message=f"Unknown container status: {container_status}")
```

Reply:
I agree, would you like me to make these changes or wait for someone else to look consider it?

Reply:
Ah no worries. I'm working on a PR to close the docker client after use. I'll just put this change there.

Reply:
part of #8357 now